doc/wargus.6
.TH WARGUS 6 "Feb 2016" "Wargus v2.4"
.SH NAME
wargus \- Warcraft II data game set for the Stratagus engine
.SH "DESCRIPTION"
Wargus is data game set for Stratagus engine. This is only launcher and wrapper
for Stratagus, and shares most of its commandline arguments.
.TP
Wargus has specific command line arguments to quickstart a single player game.
.TP
.B -c singleplayer -G map=(path-to-map),type=0,aiplayers=1,aiscripts=(wc2-land-attack|wc2-air-attack|...),[reveal]
These arguments start a single player game with one AI opponent using the
selected AI script.
.TP
.B -c singleplayer -G map=(path-to-map),type=2,aiplayers=2,aiscripts=wc2-land-attack:wc2-air-attack
These arguments start a quick machine vs machine game, pitting the land attack
vs the air attack AI. This way of running the game can be very useful for
evaluating different AIs quickly against each other, as well as for training
external AI agents that interface through some other means with the game. This
has been used (for example) to train a reinforcement learning AI to play the
game.
.TP
Wargus has specific command line arguments to start multiplayer games:
.TP
.B -c multiplayer -G server,numplayers=(number),map=(path-to-map)[,dedicated,race=(orc|human),player=(nick),fow=(1|0),reveal=(0|1),resources=(low|medium|high),units=1]
These arguments start a multiplayer server. It will directly select the passed
map and wait for numplayers players to join and become ready (including the
server), before automatically starting the game. The numplayers and map
arguments are required, the race, player nickname, fow (Fog of war), reveal
(Reveal map), resources and units options are optional. Note that if there are
spaces or other characters that must be escaped in the nickname or map path,
you must escape them (or enclose the entire argument in quotes). If the
dedicated option is passed, the server will itself only be an AI player, and
wait for numplayers clients to connect (excluding the server).
.TP
.B -c multiplayer -G client,ip=(server-ip)[,race=(orc|human),player=(nick)]
These arguments start a multiplayer game as client. It will directly join the
specified server and announce itself as ready. The ip argument is required, the
race and player nickname are optional. Note that if there are spaces or other
characters that must be escaped in the nickname, you must escape them (or
enclose the entire argument in quotes).
.TP
See \fIstratagus\fP(6) for a full list of command line arguments.
.SH "SEE ALSO"
.PD 0
.TP
\fIstratagus\fP(6), \fIwartool\fP(6), \fIpudconvert\fP(6)
.SH NOTES
Warcraft II is a registered trademark of Blizzard Entertainment.
.SH AUTHOR
Wargus launcher is written by Pali Rohár <pali.rohar@gmail.com> .
.PP
This manual page was written by Pali Rohár <pali.rohar@gmail.com> and Tim
Felgentreff <timfelgentreff@gmail.com>.


CHANGELOG
2.2.6
 - Game
   - Show error box if saving game/map failed
   - Added support for starting game with specified map from command line
   - Animate Gold Mine when workers mining
   - Handle errors when saving game failed
   - Fixed showing tips on startup

2.2.5.5
 - Game
   - Fixed loosing wood, when peasant or peon has not full wood limit
   - Fixed multiplayer network menu
   - Fixed menus for Map Editor (start editor with param -e)
   - Added support for selecting race in multiplayer game
   - Moved definitions of sound game from stratagus binary to wargus scripts
 - Pudconvert
   - Ignore unknow units when converting maps - bug #672845
 - Makefile
   - Added strip, pack and installer section
   - Use $^ instead source files in Makefile
 - Unix build.sh
   - Fixed using wildmidi
 - Debian
   - Fixed debian scripts
   - Rewrited debian/rules to use debhelper 6
 - Maemo
   - Fixed running warextract after install
   - Fixed install path for manpages
 - Windows
   - Fixed compilation on MSVC
   - Fixed console output
   - Fixed searching for Wargus in 64 bit Windows registry
 - Windows NSIS Installer
   - Fixed removing all installed lua scripts files at uninstall
   - Updated timidity binary

2.2.5.4
 - Game
   - Fixed color of orc menu button (F10) - bug #616486
   - Fixed correct cursors for races
 - Campaigns
   - Fixed orc campaign level 2
   - Fixed background music in Briefing and PictureStep
 - Debian
   - Fixed debian scrips
 - Maemo
   - Fixed crashing when starting from menu
 - Windows NSIS Installer
   - Check for correct version of Stratagus

2.2.5.3
 - Game
   - Fixed UI theme in main menu and other menus
   - Fixed ballista shooting
   - Fixed loading saved game from playing camapign game
   - Added different music playlist for orc (only orc themes) and for human (only human themes)
   - Added Briefing, Victory and Defeat music
   - Added save game confirmation menu
 - Wartool
   - Extract midi version of audio CD tracks from WAR archives
 - Wargus
   - Added support to run Wargus on Linux in console framebuffer mode
 - Unix build.sh
   - Fixed paths
   - Added support for wildmidi (alternative to timidity)
 - Debian
   - Fixed build depends
   - Fixed debian scripts
 - Maemo
   - Fixed info for Hildon Application Manager
   - Fixed debian scripts
   - Use Backspace key for enter game menu
 - Windows NSIS Installer
   - Added support for converting audio midi files (using timidity and freepats)
   - Do not extract data again after reinstalling
   - Use MUI2 instead MUI
 - Documentation
   - Rewrited README

2.2.5.2
 - Game
   - Fixed name of default map
   - Fixed WM title and icon
   - Added option for maximum OpenGL textures
   - Added version info to menu
   - Added multiplatform program wargus for searching stratagus and correct start-up
   - Rip hidden track I'm Medieval Man and add cheat disco to run this track
 - Campaigns
   - Fixed campaign triggers after loading saved game
   - Fixed paths of files in campaigns
   - Fixed speed of scrolling text in briefing
   - Added ending victory video to campaign menu
   - Added missing ending vicotry text screen at end of campaigns
   - Show next campaign level in campaign menu after completed saved game
   - Show only finished levels in campaign submenu
   - Play video files in correct order
 - Wartool
   - Fixed converting PUD maps on Unix
   - Fixed extracting videos with correct names
   - Extract ending victory campaigns texts
   - Write campaigns title and objectives directly to files without templates
 - Makefile
   - Created new Makefile, which support win32 builds and wargus starter
 - Debian
   - Fixed debian scripts
   - Rewrited config scripts for better extract game data
 - Maemo
   - Fixed debian scripts for Maemo sections
   - Fixed using findutils
   - Added program warextract for extracting data on Maemo
   - Enable using ffmpeg2theora
 - Windows
   - Fixed espace paths
   - Added missing Windows resource file for Wargus icon
 - Windows NSIS Installer
   - Fully rewrited and created new installer
   - Fixed adding campaigns maps and scripts files
   - Fixed including data files
   - Added support for checking if Stratagus is installed
   - Added support for 64 bit version
   - Added missing \r\n for string compare if data was extracted
   - Added support for extracting data
   - Added support for converting video files (using ffmpeg2theora)
   - Added support for ripping audio tracks (using cdda2wav)
   - Added desktop icon
   - Include default music file
   - Download ffmpeg2theora and cdrtools when creating installer
   - Use plugin DosExec instead function ExecWait

2.2.5.1
 - Game
   - Added support for more graphics resolutions
   - Added intro video and correct title screen
 - Campaigns
   - Added missing videos between levels
   - Added submenu for levels (with act screen and videos)
 - Wartool
   - Added support for extract data from CD if file names are upper
   - Fixed extract number of logo video
   - Fixed converting maps PUD files
   - Extract ui graphics for more graphics resolutions
   - Extract sounds for campaigns acts screen
   - Removed campaigns template files, instead create new and write title and objectives directly
   - Store version of wartool, in future do not extract data twice
 - Unix build.sh
   - Added param -d for cdparanoia CD drive
   - Fixed searching for data path
   - Fixed file names for extract
   - Check if cdparanoia and ffmpeg2theora are in system
   - Do not rip audio tracks if are ripped in data path
   - For decode audio tracks use ffmpeg2theora instead oggenc
   - Convert extracted videos from SMK format to theroa (with ffmpeg2theora)
 - Debian
   - Added missing Wargus files do deb package
 - Maemo
   - Added optification on Maemo
   - Added info for Hildon Application Manager
 - Windows
   - Removed old Windows extract data programs


wargus.cpp
/*
       _________ __                 __
      /   _____//  |_____________ _/  |______     ____  __ __  ______
      \_____  \\   __\_  __ \__  \\   __\__  \   / ___\|  |  \/  ___/
      /        \|  |  |  | \// __ \|  |  / __ \_/ /_/  >  |  /\___ |
     /_______  /|__|  |__|  (____  /__| (____  /\___  /|____//____  >
             \/                  \/          \//_____/            \/
  ______________________                           ______________________
                        T H E   W A R   B E G I N S
         Stratagus - A free fantasy real time strategy game engine

    wargus.c - Wargus Game Launcher
    Copyright (C) 2010-2011  Pali Rohár <pali.rohar@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include "wargus.h"

#define GAME_NAME "Wargus"
#define GAME_CD "Warcraft II DOS REZDAT.WAR or BNE INSTALL.MPQ/INSTALL.EXE file"
#define GAME_CD_FILE_PATTERNS "REZDAT.WAR", "rezdat.war", "War Resources", "INSTALL.EXE", "Install.exe", "Install.mpq", "INSTALL.MPQ", "install.exe", "install.mpq"
#define GAME "wargus"
#define EXTRACTOR_TOOL "wartool"
#define EXTRACTOR_ARGS "-v -r"
#define CHECK_EXTRACTED_VERSION 1

#define EXTRACTION_FILES REEXTRACT_MARKER_FILE

#define __wargus_contrib__ "campaigns", "campaigns", \
                           "contrib", "graphics/ui", \
                           "maps", "maps", \
                           "shaders", "shaders", \
                           "scripts", "scripts"
                           // ":optional:", \
                           // "music/TimGM6mb.sf2", "music/TimGM6mb.sf2"

#define CONTRIB_DIRECTORIES { __wargus_contrib__, NULL }
#define GAME_SHOULD_EXTRACT_AGAIN (tinyfd_messageBox("Extract more?", \
                                                     "Extract from an additional expansion CD?", \
                                                     "yesno",           \
                                                     "question",        \
                                                     1))

const char* SRC_PATH() { return __FILE__; }

#ifdef WIN32
#define TITLE_PNG "%s\\graphics\\ui\\title.png"
#else
#define TITLE_PNG "%s/graphics/ui/title.png"
#endif

#undef DATA_PATH
#define DATA_PATH "invalid"

#include <stratagus-game-launcher.h>
